feat: adds unit tests for sync and async reads of a pipeline transformation

kevkim-codes opened the pull request:
Discovered there may be some unit test coverage needed to test pipeline transformations during a Cloud Bigtable feature PR (https://github.com/googleapis/nodejs-bigtable/pull/1494).

kevkim-codes committed:
add unit test for pipeline transformations

kevkim-codes committed:
add a unit test for a pipeline transformation

leahecole commented on gax/test/unit/streaming.ts:
```diff
@@ -1466,6 +1474,67 @@ describe('streaming', () => {
+      setImmediate(() => {
+        s.emit('metadata');
+      });
+      return s;
```
I think you should probably also be emitting "status" after metadata to properly be imitating a grpc stream

leahecole commented on gax/test/unit/streaming.ts:
```diff
@@ -1466,6 +1474,67 @@ describe('streaming', () => {
+    });
+
+    // Initial stream.
+    const apiCall = createApiCallStreaming(
+      spy,
+      streaming.StreamType.SERVER_STREAMING
+    );
```
```suggestion
  const apiCall = createApiCallStreaming(
      spy,
      streaming.StreamType.SERVER_STREAMING,
      false,
      true // new retry behavior disabled
    );
```
gotta make sure this goes through the new retry stuff

leahecole commented on gax/test/unit/streaming.ts:
```diff
@@ -1466,6 +1474,67 @@ describe('streaming', () => {
+    const finalResults: Array<{resources: Array<number>}> = [];
+
+    s1.on('error', (err: GoogleError) => {
+      console.log('s1 error');
```
```suggestion
```

leahecole commented on gax/test/unit/streaming.ts:
```diff
@@ -1466,6 +1474,67 @@ describe('streaming', () => {
+      done();
+    });
+
+    pipeline(s1, transform, s2, () => {});
```
It would be worth adding error handling here along the lines of 
```suggestion
    pipeline(s1, transform, s2,  (err) => {
    if (err) {
      throw new Error('pipeline in properly emits the end event at the end of a pipeline transformation test failed')
    });
```

leahecole commented on gax/test/unit/streaming.ts:
```diff
@@ -1466,6 +1474,67 @@ describe('streaming', () => {
+
+    s1.on('error', (err: GoogleError) => {
+      console.log('s1 error');
+      done(err);
+    });
+
+    s2.on('error', (err: GoogleError) => {
+      done(err);
```
These "done"s should not be here. If you are expecting an error as part of this test, you should add assertions in these event handlers to make sure that the error is what you expect. If you are NOT expecting an error, please throw the error so that the test can fail

leahecole replied:
If you wanted, you could also add some kind of error counter that you increment if you hit an error and then you could assert in your .on('end') that it's 0, but that might be overkill.

leahecole reviewed (changes requested):
A couple of overall pieces of feedback:

- I think a test should also be added to the test application that does this
- I know our problem before was the test timing out because "done" was never called, but also make sure it's possible for it to fail and to fail LOUDLY :)! It might in its current state, but it might not , which is why I added those notes about "done" and the errors
-

kevkim-codes committed:
Update gax/test/unit/streaming.ts

Co-authored-by: Leah E. Cole <6719667+leahecole@users.noreply.github.com>

kevkim-codes committed:
fix unit test

kevkim-codes committed:
Merge branch 'main' of github.com:kevkim-codes/gax-nodejs

kevkim-codes committed:
incorporate review suggestions

kevkim-codes commented:
> A couple of overall pieces of feedback:
> 
> * I think a test should also be added to the test application that does this
> * I know our problem before was the test timing out because "done" was never called, but also make sure it's possible for it to fail and to fail LOUDLY :)! It might in its current state, but it might not , which is why I added those notes about "done" and the errors

1) Sounds good. I will have a separate PR for the test application.

2) Incorporated your changes in the review and made the error handling better, thank you!

kevkim-codes committed:
louder error message

kevkim-codes committed:
added an async test

leahecole commented on gax/test/unit/streaming.ts:
```diff
@@ -1466,6 +1474,138 @@ describe('streaming', () => {
+      spy,
+      streaming.StreamType.SERVER_STREAMING,
+      false,
+      true // new retry behavior disabled
```
```suggestion
      true // new retry behavior enabled
```

leahecole commented on gax/test/unit/streaming.ts:
```diff
@@ -1466,6 +1474,138 @@ describe('streaming', () => {
+      spy,
+      streaming.StreamType.SERVER_STREAMING,
+      false,
+      true // new retry behavior disabled
```
```suggestion
      true // new retry behavior enabled
```

leahecole reviewed (approved):
Just these nits

kevkim-codes committed:
Update gax/test/unit/streaming.ts

Co-authored-by: Leah E. Cole <6719667+leahecole@users.noreply.github.com>

kevkim-codes committed:
Update gax/test/unit/streaming.ts

Co-authored-by: Leah E. Cole <6719667+leahecole@users.noreply.github.com>

kevkim-codes committed:
update node version for CI

kevkim-codes committed:
change ci to only node 18

kevkim-codes committed:
unrevert node changes

kevkim-codes committed:
fix node engine

kevkim-codes committed:
fix typo

kevkim-codes committed:
change engine

kevkim-codes committed:
fix package json

kevkim-codes committed:
change node engine to 16

kevkim-codes committed:
fix engine

kevkim-codes committed:
fix engine

kevkim-codes committed:
change compodoc version to be supported by node 16

kevkim-codes committed:
override compodoc dependency issue for node 16

leahecole reviewed (dismissed):
Please add the compodoc stuff as a separate pr!

leahecole commented on gax/package.json:
```diff
@@ -64,7 +64,11 @@
     "webpack-cli": "^4.0.0"
   },
-
+  "overrides": {
+    "@compodoc/compodoc": {
+      "send": "1.0"
+    }
+  },
```
```suggestion
```

leahecole commented on gax/package.json:
```diff
@@ -26,7 +26,7 @@
     "@types/uuid": "^9.0.7",
     "@babel/plugin-proposal-private-methods": "^7.18.6",
-    "@compodoc/compodoc": "1.1.23",
+    "@compodoc/compodoc": "1.1.21",
```
```suggestion
    "@compodoc/compodoc": "1.1.23",
```

kevkim-codes committed:
fix package json issues

kevkim-codes committed:
Merge branch 'main' into main

kevkim-codes commented:
> Please add the compodoc stuff as a separate pr!
Removed!

kevkim-codes committed:
revert node engine
